docs/proposals/edge-node-tasks-design.md

```diff
@@ -0,0 +1,306 @@
+
+	// Specification of the desired behavior of NodeUpgradeJob.
+	// +optional
+	Spec NodeUpgradeJobSpec `json:"spec,omitempty"`
+	// Most recently observed status of the NodeUpgradeJob.
+	// +optional
+	Status NodeUpgradeJobStatus `json:"status,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// NodeUpgradeJobList is a list of NodeUpgradeJob.
+type NodeUpgradeJobList struct {
+	// Standard type metadata.
+	metav1.TypeMeta `json:",inline"`
+
+	// Standard list metadata.
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	// List of NodeUpgradeJobs.
+	Items []NodeUpgradeJob `json:"items"`
+}
+
+``// NodeUpgradeJobSpec is the specification of the desired behavior of the NodeUpgradeJob.
+// NodeUpgradeJobSpec is the specification of the desired behavior of the NodeUpgradeJob.
+type NodeUpgradeJobSpec struct {
+    // +Required: Version is the EdgeCore version to upgrade.
+    Version string `json:"version,omitempty"`
+    // TimeoutSeconds limits the duration of the node upgrade job.
+    // Default to 300.
+    // If set to 0, we'll use the default value 300.
+    // +optional
+    TimeoutSeconds *uint32 `json:"timeoutSeconds,omitempty"`
+    // NodeNames is a request to select some specific nodes. If it is non-empty,
+    // the upgrade job simply select these edge nodes to do upgrade operation.
+    // Please note that sets of NodeNames and LabelSelector are ORed.
+    // Users must set one and can only set one.
+    // +optional
+    NodeNames []string `json:"nodeNames,omitempty"`
+    // LabelSelector is a filter to select member clusters by labels.
+    // It must match a node's labels for the NodeUpgradeJob to be operated on that node.
+    // Please note that sets of NodeNames and LabelSelector are ORed.
+    // Users must set one and can only set one.
+    // +optional
+    LabelSelector *metav1.LabelSelector `json:"labelSelector,omitempty"`
+    // Image specifies a container image name, the image contains: keadm and edgecore.
+    // keadm is used as upgradetool, to install the new version of edgecore.
+    // The image name consists of registry hostname and repository name,
+    // if it includes the tag or digest, the tag or digest will be overwritten by Version field above.
+    // If the registry hostname is empty, docker.io will be used as default.
+    // The default image name is: kubeedge/installation-package.
+    // +optional
+    Image string `json:"image,omitempty"`
+    // Concurrency specifies the max number of edge nodes that can be upgraded at the same time.
+    // The default Concurrency value is 1.
+    // +optional
+    Concurrency int32 `json:"concurrency,omitempty"`
+    
+    // CheckItems specifies the items need to be checked before the task is executed.
+    // The default CheckItems value is nil.
+    // +optional
+    CheckItems []string `json:"checkItems,omitempty"`
+
+    FailureTolerate float64 `json:"failureTolerate,omitempty"`
```

Comment:
Why does the design use the float type here, but the actual code use the string type?

Reply:
I've updated the proposal.

Reply:
I think a float type would be more appropriate, otherwise you have to determine if the failureTolerate is empty when converting to a floating point number.

Reply:
Yes.  Like this
```
failureTolerate:
        type: number
        format: float64
```